File: Libraries/python/beagle_linux_sdk/src/beagle_linux_sdk/button.py
import gpiod
from .pin import Pin


class Button:
    def __init__(self, pin: Pin) -> None:
        self.pin = pin.gpio_input()

    def wait_for_press(self):
        while True:
            if self.pin.event_wait():
                evt = self.pin.event_read()
                if evt.type == gpiod.LineEvent.FALLING_EDGE:
                    return

    def wait_for_release(self):
        while True:
            if self.pin.event_wait():
                evt = self.pin.event_read()
                if evt.type == gpiod.LineEvent.RISING_EDGE:
                    return

File: Libraries/rust/beagle_helper/src/devices/button.rs
pub struct Button(gpiod::Lines<gpiod::Input>);

impl Button {
    pub fn new(pin: crate::pin::Pin) -> std::io::Result<Self> {
        pin.gpio_input().map(Self)
    }

    pub fn wait_for_press(&mut self) -> std::io::Result<()> {
        loop {
            let evt = self.0.read_event()?;
            if evt.edge == gpiod::Edge::Falling {
                return Ok(());
            }
        }
    }

    pub fn wait_for_release(&mut self) -> std::io::Result<()> {
        loop {
            let evt = self.0.read_event()?;
            if evt.edge == gpiod::Edge::Rising {
                return Ok(());
            }
        }
    }
}
